ktor-server/ktor-server-plugins/ktor-server-auth-jwt/jvm/src/io/ktor/server/auth/jwt/JWTAuth.kt

```diff
@@ -326,60 +353,37 @@ public class JWTAuthenticationProvider internal constructor(config: Configuratio
 /**
  * Installs JWT Authentication mechanism
  */
-public fun Authentication.Configuration.jwt(
+public fun AuthenticationConfig.jwt(
     name: String? = null,
-    configure: JWTAuthenticationProvider.Configuration.() -> Unit
+    configure: JWTAuthenticationProvider.Config.() -> Unit
 ) {
-    val provider = JWTAuthenticationProvider.Configuration(name).apply(configure).build()
-    val realm = provider.realm
-    val authenticate = provider.authenticationFunction
-    val verifier = provider.verifier
-    val schemes = provider.schemes
-    provider.pipeline.intercept(AuthenticationPipeline.RequestAuthentication) { context ->
-        val token = provider.authHeader(call)
-        if (token == null) {
-            context.bearerChallenge(AuthenticationFailedCause.NoCredentials, realm, schemes, provider.challengeFunction)
-            return@intercept
-        }
-
-        try {
-            val principal = verifyAndValidate(call, verifier(token), token, schemes, authenticate)
-            if (principal != null) {
-                context.principal(principal)
-                return@intercept
-            }
-
-            context.bearerChallenge(
-                AuthenticationFailedCause.InvalidCredentials,
-                realm,
-                schemes,
-                provider.challengeFunction
-            )
-        } catch (cause: Throwable) {
-            val message = cause.message ?: cause.javaClass.simpleName
-            JWTLogger.trace("JWT verification failed: {}", message)
-            context.error(JWTAuthKey, AuthenticationFailedCause.Error(message))
-        }
-    }
+    val provider = JWTAuthenticationProvider.Config(name).apply(configure).build()
     register(provider)
 }
 
+/**
+ * A context for [JWTAuthChallengeFunction]
+ */
+public class JwtChallengeContext(
+    public val call: ApplicationCall
+)
+
 /**
  * Specifies what to send back if session authentication fails.
  */
 public typealias JWTAuthChallengeFunction =
-    suspend PipelineContext<*, ApplicationCall>.(defaultScheme: String, realm: String) -> Unit
+    suspend JwtChallengeContext.(defaultScheme: String, realm: String) -> Unit
 
 private fun AuthenticationContext.bearerChallenge(
```

Comment:
Consider extracting public API to the separate file to make it easier to inspect

Reply:
To keep source compatibility and have `call: ApplicationCall` as a receiver